test/e2e/set_delete_test.go

```diff
@@ -0,0 +1,252 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"k8s.io/client-go/kubernetes"
+)
+
+var podCount int
+
+func testHeaders(t *testing.T, image string, route *routev1.Route, address string, headers []string, expectedResponse string, expectedMatches int) {
+	t.Helper()
+
+	kubeConfig, err := config.GetConfig()
+	if err != nil {
+		t.Fatalf("failed to get kube config: %v", err)
+	}
+	client, err := kubernetes.NewForConfig(kubeConfig)
+	if err != nil {
+		t.Fatalf("failed to create kube client: %v", err)
+	}
+
+	var extraCurlArgs []string
+	for _, header := range headers {
+		extraCurlArgs = append(extraCurlArgs, "-H", header)
+	}
+	extraCurlArgs = append(extraCurlArgs, "-v", "--resolve", route.Spec.Host+":80:"+address)
+	podCount++
+	name := fmt.Sprintf("%s%d", route.Name, podCount)
+	clientPod := buildCurlPod(name, route.Namespace, image, route.Spec.Host, address, extraCurlArgs...)
+	if err := kclient.Create(context.TODO(), clientPod); err != nil {
+		t.Fatalf("failed to create pod %s/%s: %v", clientPod.Namespace, clientPod.Name, err)
+	}
+	defer func() {
+		if err := kclient.Delete(context.TODO(), clientPod); err != nil {
+			if !errors.IsNotFound(err) {
+				t.Fatalf("failed to delete pod %s/%s: %v", clientPod.Namespace, clientPod.Name, err)
+			}
+		}
+	}()
+	expectedResponse = strings.ToLower(expectedResponse)
+	err = wait.PollImmediate(1*time.Second, 4*time.Minute, func() (bool, error) {
+		readCloser, err := client.CoreV1().Pods(clientPod.Namespace).GetLogs(clientPod.Name, &corev1.PodLogOptions{
+			Container: "curl",
+			Follow:    false,
+		}).Stream(context.TODO())
+		if err != nil {
+			t.Logf("failed to read output from pod %s: %v", clientPod.Name, err)
+			return false, nil
+		}
+		scanner := bufio.NewScanner(readCloser)
+		defer func() {
+			if err := readCloser.Close(); err != nil {
+				t.Errorf("failed to close reader for pod %s: %v", clientPod.Name, err)
+			}
+		}()
+		var numMatches int
+		for scanner.Scan() {
+			line := scanner.Text()
+			if strings.Contains(strings.ToLower(line), expectedResponse) {
+				numMatches++
+				t.Logf("found match %d of %d expected: %s", numMatches, expectedMatches, line)
+			}
+		}
+		if numMatches > 0 && numMatches != expectedMatches {
+			return false, fmt.Errorf("got %d matches for %q, expected %d", numMatches, expectedResponse, expectedMatches)
+		}
+		return numMatches == expectedMatches, nil
+	})
+	if err != nil {
+		pod := &corev1.Pod{}
+		podName := types.NamespacedName{
+			Namespace: clientPod.Namespace,
+			Name:      clientPod.Name,
+		}
+		if err := kclient.Get(context.TODO(), podName, pod); err != nil {
+			t.Errorf("failed to get pod %s: %v", clientPod.Name, err)
+		}
+
+		logs, err := client.CoreV1().Pods(clientPod.Namespace).GetLogs(clientPod.Name, &corev1.PodLogOptions{
+			Container: "curl",
+			Follow:    false,
+		}).DoRaw(context.TODO())
+		if err != nil {
+			t.Errorf("failed to get logs from pod %s: %v", clientPod.Name, err)
+		}
```

Comment:
I don't think you need to get the pod and logs again here.

Reply:
when it fails we need to show the logs which will show the user  that the desired header was not present in the logs.

Reply:
It's too much data in the test output.  The test output should be brief.  Have confidence in the test to make a correct assessment.  This feels like a debug statement for the test that you would use while it's under development.

Reply:
We wanted to put it back in, so you can resolve this comment.